File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import defaultdict
from collections import deque
def solution(n, roads, sources, d):
    ans = []
    # 노드 간의 비용 고정 => bfs
    graph = defaultdict(list)
    for a,b in roads:
        graph[a].append(b)
        graph[b].append(a)
    dis = [-1] * (n+1)
    dis[d] = 0
    visit = [False]*(n+1)
    visit[d] = True
    q = deque()
    q.append((d,0))
    while q:
        cur,time = q.popleft()
        for nxt in graph[cur]:
            if not visit[nxt]:
                visit[nxt] = True
                dis[nxt] = time + 1
                q.append((nxt,time+1))
    for s in sources:
        ans.append(dis[s])
    return ans

File: 프로그래머스/3/132266. 부대복귀/부대복귀.kt
import java.util.*
class Solution {
    fun solution(n: Int, roads: Array<IntArray>, sources: IntArray, d: Int): IntArray {
        val graph = Array(n+1){ mutableListOf<Int>() }
        for (r in roads) {
            val (a,b) = r
            graph[a].add(b)
            graph[b].add(a)
        }
        val dis = IntArray(n+1){-1}
        val q: Queue<Int> = LinkedList()
        dis[d] = 0
        q.add(d)
        while (q.isNotEmpty()) {
            val cur = q.poll()
            for (next in graph[cur]) {
                if (dis[next] == -1) {
                    dis[next] = dis[cur] + 1
                    q.add(next)
                }
            }
        }
        return sources.map { dis[it] }.toIntArray()
    }
}